doc/pandas-2.0/source/internal-architecture.rst

```diff
@@ -0,0 +1,714 @@
+
+* **Logical data type**: metadata which describes the semantic content of a
+  single value in an array or other collection of values. Depending on the
+  logical type, it may map 1-to-1 to a physical type or not at all. Here are
+  some examples:
+
+  * The ``double`` or ``float64`` type may be viewed both as a logical type as
+    well as a physical type (a 1-to-1 correspondence).
+
+  * pandas's ``category`` dtype contains its own auxiliary array of category
+    values (for example, the distinct strings collected from a string
+    array). Based on the number of categories, the category ``codes`` (which
+    reference the categories array) are stored in the smallest possible integer
+    physical type (from ``int8`` to ``int64``, depending whether the data type
+    can accommodate the codes). For example, if there are 50 codes, the data is
+    represented in ``int8`` storage. For 1000 codes, it would be ``int16``.
+
+  * Another example: timestamps may be physically stored in ``int64``
+    storage, and these values are interpreted in the context of a particular
+    time unit or resolution (e.g. nanoseconds, milliseconds, seconds).
+
+In general, new logical types may be formed either by placing new semantics on
+top of a single physical data type or some composition of physical or logical
+types. For example: you could have a categorical type (a logical construct
+consisting of multiple arrays of data) whose categories are some other logical
+type.
+
+For historical reasons, **pandas never developed a clear or clean semantic
+separation in its user API between logical and physical data types**. Also, the
+addition of new, pandas-only "synthetic" dtypes that are unknown to NumPy (like
+categorical, datetimetz, etc.) has expanded this conflation considerably. If
+you also consider pandas's custom missing / NULL data behavior, the addition of
+ad hoc missing data semantics to a physical NumPy data type created, by the
+definitions above, a logical data type (call it ``object[nullable]`` for an
+object array) without ever explicitly saying so.
+
+You might be thinking, "Good job, Wes. You really messed that up!" I'd be
+inclined to agree with you now in retrospect, but back in 2011 pandas was not
+the super popular project that it is today, and we were truly riding on NumPy's
+coat tails. So the extent to which NumPy concepts and APIs were used explicitly
+in pandas made the library easier to adopt. Now in 2016, this feels
+anachronistic / outdated.
+
+High-level logical type proposal
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As we have been discussing periodically on the pandas-dev mailing list and
+GitHub, I am proposing that we start to unravel our current mess by defining
+pandas-specific metadata objects that model the current semantics / behavior of
+the project. What does this mean, exactly?
+
+* Each NumPy dtype object will map 1-to-1 to an equivalent ``pandas.DataType``
```

Comment:
I can write down a hierarchy. Any `pandas.DataType` is a *logical type* (perhaps we can come up with a better name for this base class?)  — all of the pandas metadata objects are logical, with a clear default mapping onto a physical memory representation. 

For example:

- `pandas.Int64Type` —> `numpy.int64` plus a pandas internal bitmap
- `pandas.Float64Type` —> `numpy.float64`
- `pandas.CategoricalType` —> one of `numpy.int8` to `numpy.int64` depending on the categories
- `pandas.StringType` —> dictionary-encoded UTF8 representation described in the docs

and so on. I'll add it to the document — the main point is that there's a 1-1 mapping from NumPy's physical types onto pandas's logical types, but the mapping from pandas to NumPy may not by 1 to 1, and may not map onto NumPy at all without an explicit lossy cast (instead of the implicit lossy casts happening right now: the example cited in the docs is missing data in integer arrays).